diem-move/parallel-executor/src/scheduler.rs

```diff
@@ -1,113 +1,470 @@
-            // we want to make things fast inside the lock, so use take instead of clone
-            let mut stored_deps = self.txn_dependency[version].write().unwrap();
+    // After txn is executed, schedule its dependencies for re-execution. In some cases,
+    // may return a validation task of the transaction to the caller (otherwise NoTask).
+    pub fn finish_execution<'a>(
+        &self,
+        txn_idx: TxnIndex,
+        incarnation: Incarnation,
+        revalidate_suffix: bool,
+        guard: TaskGuard<'a>,
+    ) -> SchedulerTask<'a> {
+        self.set_executed_status(txn_idx, incarnation);
+
+        let txn_deps: Vec<TxnIndex> = {
+            let mut stored_deps = self.txn_dependency[txn_idx]
+                .lock()
+                .expect("Dependencies must exist");
+            // Holding the lock, take dependency vector.
             std::mem::take(&mut stored_deps)
         };
 
-        version_deps.sort_unstable();
-        for dep in version_deps {
-            self.txn_buffer.push(dep);
+        let mut min_dep = self.block_size;
+        for dep in txn_deps {
+            // Mark the status of dependencies as 'ReadyToExecute' since dependency on
+            // transaction txn_idx is now resolved.
+            self.resume(dep);
+
+            // Find the minimum index among resolved dependencies.
+            min_dep = min(min_dep, dep);
+        }
+        if min_dep < self.block_size {
+            // Decrease the execution index as necessary to ensure resolved dependencies
+            // get a chance to be re-executed.
+            self.decrease_execution_idx(min_dep);
+        }
+
+        // If validation_idx is already lower than txn_idx, all required transactions will be
+        // considered for validation, and there is nothing to do.
+        if self.validation_idx.load(Ordering::SeqCst) > txn_idx {
+            if revalidate_suffix {
+                // The transaction execution required revalidating all higher txns (not
+                // only itself), currently happens when incarnation writes to a new path
+                // (w.r.t. the write-set of its previous completed incarnation).
+                self.decrease_validation_idx(txn_idx);
+            } else {
+                // Only transaction txn_idx requires validation. Return validation task
+                // back to the caller. No need to change active tasks (-1 +1= 0)
+                return SchedulerTask::ValidationTask((txn_idx, incarnation), guard);
+            }
+        }
+
+        SchedulerTask::NoTask
+    }
+
+    // Finalize a validation task of version (txn_idx, incarnation). In some cases,
+    // may return a re-execution task back to the caller (otherwise, NoTask).
+    pub fn finish_validation<'a>(
+        &self,
+        txn_idx: TxnIndex,
+        incarnation: Incarnation,
+        aborted: bool,
+        guard: TaskGuard<'a>,
+    ) -> SchedulerTask<'a> {
+        if aborted {
```

Comment:
I think we can move the check out, the caller already checks this, so the function becomes `finish_abort`

Reply:
same here, benefits of wrapping num_active_tasks!!